src/boot/efi/boot.c

```diff
@@ -1683,10 +1683,10 @@ static INTN config_entry_compare(const ConfigEntry *a, const ConfigEntry *b) {
                         return r;
         }
 
-        /* Now order by ID (the version is likely part of the ID, thus note that this might put the oldest
-         * version last, not first, i.e. specifying a sort key explicitly is thus generally preferable, to
-         * take benefit of the explicit sorting above.) */
-        r = strverscmp_improved(a->id, b->id);
+        /* Now order by ID. The version is likely part of the ID, thus note that this will generatelly put
+         * the newer versions earlier. Specifying a sort key explicitly is preferable, because it gives an
+         * explicit sort order. */
+        r = -strverscmp_improved(a->id, b->id);
```

Comment:
i don't get this?

before arch would be placed before fedora. But now fedora before arch? 

the `id` field is a concatenation of os name and version typically. This means you can't have both when using strverscmp(): either os names *and* versions are ordered in increasing order, or they are ordered in decreasing order. You can't have os names up but versions down...

The thing is that there is no clear naming regime how the ids are picked, and hence it's hard to reconstruct what is os name and what is os version

Reply:
`id` is actually the filename. And it is usually built as `<machine-id>-<version>.conf`. In that case, the `<machine-id>` part is an opaque string and it does not matter whether we sort them in increasing or decreasing order. But within a single machine-id group, we sort by `<version>`, and this is the point of the patch: within a group of entries with the same machine-id, entries are sorted with newest at the top. I think this is pretty important, because a) non-dual-boot machines are a big majority, b) people have been confused by reversed sort order and it has been a long-standing issue to sort newest-at-the-top, c) this sort order is consistent with the "new" sort order for entries with sort-key.

It is true that the filename might be constructed differently, or that we might sort groups of entries non-alphabetically. But we already did that, e.g. putting auto-windows or macos last. In fact even with sort-key, the entry titles might be sorted alphabetically or not. And as mentioned above, dual-boot machines are the minority. And dual-boot machines where at least two systems are not windows are even rarer. So overall, this patch should make things better for the common case, and is "meh" for the other cases.

For example:
on my machine I have:
`/boot/efi/loader/entries/08a5690a2eed47cf92ac0a5d2e3cf6b0-5.15.14-200.fc35.x86_64.conf`
`/boot/efi/loader/entries/08a5690a2eed47cf92ac0a5d2e3cf6b0-5.17.0-0.rc5.102.fc36.x86_64.conf`
and this is parsed into:
`id = 08a5690a2eed47cf92ac0a5d2e3cf6b0-5.15.14-200.fc35.x86_64.conf`
and `id = 08a5690a2eed47cf92ac0a5d2e3cf6b0-5.17.0-0.rc5.102.fc36.x86_64.conf`
and results in 5.17.0 being sorted as newer then 5.15.14, which is all I care about as a user.

If I add a second system, and I have
`id = 99abbbba2eed47cf92ac0a5d2e3cf6b0-5.1.14-200.fc29.x86_64.conf`
and `id = 99abbbba2eed47cf92ac0a5d2e3cf6b0-4.19.0-0.fc29.x86_64.conf`,
I care that `5.1.14` is above `4.19.0`. Whether this group is above or below the other one isn't really important. Or in other words, the sort order of `99abbbba2eed47cf92ac0a5d2e3cf6b0` and `08a5690a2eed47cf92ac0a5d2e3cf6b0` doesn't matter, as long as it's stable.